### PR TITLE
[neox] consensus

### PR DESCRIPTION
**Pull request by KickSeason**
### mpt integration (Step 1/4) 
Process state root in consensus.
* Add `StateRootSignature` when send `PrepareRequest`
* Check `StateRootSignature` proposaled in `PrepareRequest` received.
* Relay `StateRoot` when commit.

@erikzhang @shargon @Tommo-L Can you please have a look?

### CODE REVIEW

**Comment by KickSeason** [comment]
We have another option is add state root signature in  `PreparationPayloads` and relay `StateRoot` when commit.

In testnet, We find that if only one cn collected all commits and relay block but doesn't relay state root, the whole network will lose state root.

File: src/neo/Consensus/Commit.cs [hunk 1]
```diff
@@ -5,22 +5,24 @@ namespace Neo.Consensus
 {
     public class Commit : ConsensusMessage
     {
-        public byte[] Signature;
-
-        public override int Size => base.Size + Signature.Length;
+        public byte[] BlockSignature;
+        public byte[] StateRootSignature;
```
Comment (erikzhang) [inline src/neo/Consensus/Commit.cs:9]: This should be the state root of the previous block.

Comment (KickSeason) [reply src/neo/Consensus/Commit.cs:9]: Put state root of previous block in `commit` and relay state root with one signature when collected enough `commit`?

File: src/neo/Ledger/Blockchain.state.cs [hunk 2]
```diff
@@ -0,0 +1,12 @@
+using Akka.Actor;
+
+namespace Neo.Ledger
+{
+    public sealed partial class Blockchain : UntypedActor
+    {
+        public UInt256 GetLocalStateRoot(uint index)
+        {
+            return UInt256.Zero;
```
Comment (shargon) [inline src/neo/Ledger/Blockchain.state.cs:9]: It will be filled in other PR?

File: src/neo/Network/P2P/Payloads/StateRoot.cs [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+        {
+            return this.GetHashData();
+        }
+
+        public JObject ToJson()
+        {
+            var json = new JObject();
+            json["version"] = Version;
+            json["index"] = Index;
+            json["stateroot"] = Root.ToString();
```
Comment (shargon) [inline src/neo/Network/P2P/Payloads/StateRoot.cs:121]: `ToJson` it's required?

File: src/neo/Network/P2P/Payloads/StateRoot.cs [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+                return new[] { Witness };
+            }
+            set
+            {
+                if (value.Length != 1) throw new ArgumentException();
+                Witness = value[0];
+            }
+        }
+
+        public int Size =>
```
Comment (shargon) [inline src/neo/Network/P2P/Payloads/StateRoot.cs:49]: Add comments of each value

File: src/neo/Network/P2P/Payloads/StateRoot.cs [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+
+        public bool Verify(StoreView snapshot)
+        {
+            return this.VerifyWitnesses(snapshot, 1_00000000);
+        }
+
+        public virtual UInt160[] GetScriptHashesForVerifying(StoreView snapshot)
+        {
+            var script_hash = Blockchain.Singleton.GetBlock(Index)?.NextConsensus;
+            return script_hash is null ? Array.Empty<UInt160>() : new UInt160[] { script_hash };
```
Comment (shargon) [inline src/neo/Network/P2P/Payloads/StateRoot.cs:108]: if `script_hash` is null we should throw an exception?

File: src/neo/Network/P2P/Payloads/StateRoot.cs [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+        }
+
+        public byte[] GetMessage()
+        {
+            return this.GetHashData();
```
Comment (shargon) [inline src/neo/Network/P2P/Payloads/StateRoot.cs:113]: Where is this used?

File: src/neo/Ledger/Blockchain.state.cs (hunk 2 above)
Comment (KickSeason) [reply src/neo/Ledger/Blockchain.state.cs:9]: Yes. My changes in consensus need interface in Blockchain. And blockchain changes need persistence changes merged. I have no choice.

File: src/neo/Network/P2P/Payloads/StateRoot.cs (hunk 3 above)
Comment (KickSeason) [reply src/neo/Network/P2P/Payloads/StateRoot.cs:121]: We might need return this in RPC.

Comment (shargon) [reply src/neo/Network/P2P/Payloads/StateRoot.cs:121]: RPC it's a plugin, I think that we should remove it from core

Comment (KickSeason) [reply src/neo/Network/P2P/Payloads/StateRoot.cs:116]: Why we keep `Block.ToJson()`

File: src/neo/Network/P2P/Payloads/StateRoot.cs (hunk 6 above)
Comment (KickSeason) [reply src/neo/Network/P2P/Payloads/StateRoot.cs]: removed.

File: src/neo/Consensus/Commit.cs (hunk 1 above)
Comment (vncoelho) [reply src/neo/Consensus/Commit.cs:9]: The state root of previous block may need to follow a full consensus agreement with `2F + 1`, perhaps we need to send that on `PrepResp`and `PrepReq`

File: src/neo/Consensus/ConsensusContext.cs [hunk 7]
```diff
@@ -24,6 +26,8 @@ internal class ConsensusContext : IDisposable, ISerializable
         private const byte ConsensusStatePrefix = 0xf4;
 
         public Block Block;
+        public UInt256 ProposalStateRoot { get; set; }
```
Comment (vncoelho) [inline src/neo/Consensus/ConsensusContext.cs:29]: There should not be a `ProposalStateRoot`,
we have `PreviousBlockStateRoot` and `CurrentStateRoot. The latter is relayed as a message.

File: src/neo/Consensus/Commit.cs (hunk 1 above)
Comment (KickSeason) [reply src/neo/Consensus/Commit.cs:9]: @erikzhang

Comment (erikzhang) [reply src/neo/Consensus/Commit.cs:9]: I think we can put the signatures of previous  state root into `PrepareRequest` and `PrepareResponse`. The state root can be calculated, we don't need to send it.

Comment (KickSeason) [reply src/neo/Consensus/Commit.cs:9]: Will do.

**Comment by KickSeason** [comment]
@erikzhang @shargon Can you please have a look.

**Comment by shargon** [comment]
Change to https://github.com/neo-project/neo/tree/neox

**Comment by KickSeason** [comment]
Close this to merge to ` neox` branch.

File: src/neo/Consensus/Commit.cs (hunk 1 above)
Comment (KickSeason) [reply src/neo/Consensus/Commit.cs]: #1719

**Pull request closed without merging**